Show the expected glyph in the dictionary lookup window

nairyosangha opened the pull request:
I noticed that when reading a book in japanese, the dictionary display lookup window displays the chinese glyphs.


This PR lets the user add a custom `lang=<from>-<to>` line to the dictionary .ifo file, which can be used to determine which glyphs should be shown.

before: (the selected word in the book is what it should look like)
![image](https://user-images.githubusercontent.com/34285115/223243385-7088ea1e-3cf6-48d5-8217-a0fa9c2ff563.png)

after: I've added `lang=ja-ja` to the dictionary's .ifo
![image](https://user-images.githubusercontent.com/34285115/223243857-cad6f4fb-8383-47b4-a8e6-35ed84445630.png)

There are some more things which I could tackle as well, if deemed okay:
- for HTML dictionaries, the little `(query: ..)` part at the end still shows the incorrect glyphs
![image](https://user-images.githubusercontent.com/34285115/223244560-2633f98d-d682-411f-97b8-4d6633e7466b.png)

- The title of the dictionary itself on the first picture also uses chinese glyphs, I suppose it could look at the 'from' language in the .ifo as well

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/koreader/koreader/10184)
<!-- Reviewable:end -->

poire-z commented:
Some previous discussion on gitter started at https://app.gitter.im/#/room/#koreader_koreader:gitter.im/$SFYk2eB2flFrgtx7HAcA-cEhqsqmFIjVXz2MWVojP70

> for HTML dictionaries, the little (query: ..) part at the end still shows the incorrect glyphs

How would you do that ? Does `<span lang=>` work with MuPDF ? (otherwise, if it's adding another widget in the layout, probably not worth the complexity...)

> The title of the dictionary itself on the first picture also uses chinese glyphs, I suppose it could look at the 'from' language in the .ifo as well

Sure "from" is always better than "to" ?

(The next natural step from there would be for our "download directories" feature to hack the ifo it downloads to add the lang to it (as we know them in frontend/ui/data/dictionaries, alas not in the same format).)

poire-z commented on frontend/apps/reader/modules/readerdictionary.lua:
```diff
@@ -136,6 +136,7 @@ function ReaderDictionary:init()
                 f:close()
                 local dictname = content:match("\nbookname=(.-)\r?\n")
                 local is_html = content:find("sametypesequence=h", 1, true) ~= nil
+                local lang_from, lang_to, _, _ = content:match("lang=(%a+)-?(%a*)\r?\n?")
```
you can remove the `_, _` (would they be filled ? and if they were, it's fine if there's no var to receive them in Lua.)

nairyosangha replied:
I was thinking of string.find, which returns the indexes of the match as well. Not that it really makes sense here :sweat_smile: 
I've removed it

nairyosangha commented:
> > for HTML dictionaries, the little (query: ..) part at the end still shows the incorrect glyphs
> 
> How would you do that ? Does `<span lang=>` work with MuPDF ? (otherwise, if it's adding another widget in the layout, probably not worth the complexity...)

Oh never mind, I misunderstood how that actually works, it has nothing to do with the HTML dict, it just shows it on the first dictionary in the list (a HTML dict in my case), so if that would happen to be a text dictionary, everything would be fine as is.

> 
> > The title of the dictionary itself on the first picture also uses chinese glyphs, I suppose it could look at the 'from' language in the .ifo as well
> 
> Sure "from" is always better than "to" ?
> 

Good point, if the target language is English, the title will likely be in English. Likewise if it's Japanese. It's not gonna be 100% though, but good enough I guess.
e.g. a Japanese-English dictionary called "日本語-English dictionary", which would still get the wrong glyphs if we look at the target language.

nairyosangha commented:
> (The next natural step from there would be for our "download dictionaries" feature to hack the ifo it downloads to add the lang to it (as we know them in frontend/ui/data/dictionaries, alas not in the same format).)

I've tried implementing that, but I don't have a decent way of getting from the full language (e.g. "English") to the locale which I need to store in the .ifo..

I made an attempt by using the mapping in `ui/language.lua` but that's incomplete, and it's localized so I can't use it to compare reliably.
e.g. in that file "German" is listed as "Deutsch"

Another option is to look at that big 'LANGUAGES' list in `apps/reader/modules/readertypography.lua` but that list is 
not public, and it also has the full language names in localized form (that's what calling gettext does right?).
So I would have to make that list public, and in `ui/data/dictionaries` I'd have to localize all the languages as well.

There has to be a better way :slightly_smiling_face:

nairyosangha commented:
I also noticed that the terms 'lang_in' and 'lang_out' are used in `ui/data/dictonaries.lua`. Might be nice if I use the same terminology?

poire-z commented:
> There has to be a better way 

Well, you could just add them manually once and for all into the frontend/ui/data/dictionaries.lua items, ie.:
```diff
   {
        name = "English-Russian economic dictionary",
        lang_in = "English",
        lang_out = "Russian",
+       ifo_lang = "en-ru",
        entries = 14436,
```
So we could then have a function to "copy" all those prefixed with `ifo_` into the .ifo (without that prefix).

> I also noticed that the terms 'lang_in' and 'lang_out' are used in ui/data/dictonaries.lua. Might be nice if I use the same terminology?

As you feel (no strong opinion either way).

nairyosangha commented:
Well updating that mapping took forever, I got most of these from https://en.wikipedia.org/wiki/List_of_ISO_639-1_codes, provided the list below for reference

```lua
local mapper = {
    English = 'en',
    Swedish = 'sv',
    French = 'fr',
    ["Ancient Greek"] = 'grc',
    German = 'de',
    Latin = 'la',
    Portuguese = 'pt',
    Czech = 'cs',
    Afrikaans = 'af',
    Chinese = 'zh',
    Russian = 'ru',
    Spanish = 'es',
    Danish = 'da',
    Dutch = 'nl',
    Arabic = 'ar',
    Belarusian = 'be',
    Bulgarian = 'bg',
    Finnish = 'fi',
    Hungarian = 'hu',
    Serbian = 'sr',
    Esperanto = 'eo',
    Estonian = 'et',
    Icelandic = 'is',
    Italian = 'it',
    Japanese = 'ja',
    Norwegian = 'no',
    Swahili = 'sw',
    Ukrainian = 'uk',
    Polish = 'pl',
    Adyghe = 'ady', -- from https://en.wikipedia.org/wiki/Adyghe_language
    Afar = 'aa',
    Albanian = 'sq',
    ["Alemannic German"] = 'gsw', -- from https://en.wikipedia.org/wiki/Alemannic_German
    Aramaic = 'arc', -- from https://www.loc.gov/standards/iso639-2/php/code_list.php
    Armenian = 'hy',
    Aromanian = 'rup', -- from https://en.wikipedia.org/wiki/Aromanian_language
    Assamese = 'as',
    ["Assyrian Neo-Aramaic"] = 'arc', -- from https://www.loc.gov/standards/iso639-2/php/code_list.php
    Asturian = 'ast', -- from https://www.loc.gov/standards/iso639-2/php/code_list.php
    Azerbaijani = 'az',
    Bashkir = 'ba',
    Basque = 'eu',
    Bengali = 'bn',
    ["Bikol Central"] = 'bik', -- from https://www.loc.gov/standards/iso639-2/php/code_list.php
    Breton = 'br',
    Burmese = 'my',
    Cantonese = 'yue', -- from https://en.wikipedia.org/wiki/Cantonese
    Cebuano = 'ceb', -- from https://www.loc.gov/standards/iso639-2/php/code_list.php
    ["Central Nahuatl"] = 'nah', -- https://www.loc.gov/standards/iso639-2/php/code_list.php
    Cherokee = 'chr', -- from https://www.loc.gov/standards/iso639-2/php/code_list.php
    Cimbrian = 'cim', -- from https://en.wikipedia.org/wiki/Cimbrian_language
    ["Classical Nahuatl"] = 'nah', -- https://www.loc.gov/standards/iso639-2/php/code_list.php
    ["Classical Syriac"] = 'syc', -- https://www.loc.gov/standards/iso639-2/php/code_list.php
    Coptic = 'cop', -- from https://www.loc.gov/standards/iso639-2/php/code_list.php
    Cornish = 'cor', -- from https://www.loc.gov/standards/iso639-2/php/code_list.php
    ["Crimean Tatar"] = 'crh', -- from https://www.loc.gov/standards/iso639-2/php/code_list.php
    Dalmatian = 'dlm', -- what the fuck https://en.wikipedia.org/wiki/Dalmatian_language
    Dhivehi = 'div', -- from https://www.loc.gov/standards/iso639-2/php/code_list.php
    Faroese = 'fo',
    Friulian = 'fur', -- from https://www.loc.gov/standards/iso639-2/php/code_list.php
    Galician = 'glg', -- from https://www.loc.gov/standards/iso639-2/php/code_list.php
    Georgian = 'ka',
    Gothic = 'got',
    Greek = 'el',
    Greenlandic = 'kl',
    Gujarati = 'gu',
    ["Haitian Creole"] = 'ht',
    Hausa = 'ha',
    Hawaiian = 'haw', -- from https://www.loc.gov/standards/iso639-2/php/code_list.php
    Hebrew = 'he',
    Hiligaynon = 'hil', -- from https://www.loc.gov/standards/iso639-2/php/code_list.php
    Hindi = 'hi',
    Hunsrik = 'hrx', -- from https://en.wikipedia.org/wiki/Hunsrik
    Ido = 'io',
    Indonesian = 'id',
    Ingrian = 'izh', -- from https://en.wikipedia.org/wiki/Ingrian_language
    Interlingua = 'ia',
    Irish = 'ga',
    Javanese = 'jv',
    Jeju = 'jje', -- from https://en.wikipedia.org/wiki/Jeju_language
    Kabuverdianu = 'kea', -- https://en.wikipedia.org/wiki/Cape_Verdean_Creole
    Kannada = 'kn',
    Kashmiri = 'ks',
    Kazakh = 'kk',
    Khmer = 'km',
    Kikuyu = 'ki',
    Korean = 'ko',
    Kumyk = 'kum',  -- from https://www.loc.gov/standards/iso639-2/php/code_list.php
    Kyrgyz = 'ky',
    Ladin = 'lld', -- from https://en.wikipedia.org/wiki/Ladin_language
    Ladino = 'lad', -- from https://www.loc.gov/standards/iso639-2/php/code_list.php
    Lao = 'lo',
    Latvian = 'lv',
    Laz = 'lzz', -- from https://en.wikipedia.org/wiki/Laz_language
    Lithuanian = 'lt',
    ["Lower Sorbian"] = 'wen', -- from https://www.loc.gov/standards/iso639-2/php/code_list.php
    Luxembourgish = 'lb',
    Macedonian = 'mk',
    Malagasy = 'mg',
    Malay = 'ms',
    Malayalam = 'ml',
    Maltese = 'mt',
    Manchu = 'mnc', -- from https://www.loc.gov/standards/iso639-2/php/code_list.php
    Mandarin = 'cmn', -- from https://en.wikipedia.org/wiki/Mandarin_Chinese
    Manx = 'gv',
    Maori = 'mi',
    Maquiritari = 'mch', -- from https://en.wikipedia.org/wiki/Ye%27kuana_language
    Marathi = 'mr',
    Marshallese = 'mh',
    ["Middle Dutch"] = 'dum', -- from https://www.loc.gov/standards/iso639-2/php/code_list.php
    ["Middle English"] = 'enm', -- from https://www.loc.gov/standards/iso639-2/php/code_list.php
    ["Middle French"] = 'frm', -- from https://www.loc.gov/standards/iso639-2/php/code_list.php
    Mongolian = 'mn',
    ["Moroccan Arabic"] = 'ary', -- from https://en.wikipedia.org/wiki/Moroccan_Arabic
    Navajo = 'nv',
    Norman = 'nrf', -- from https://en.wikipedia.org/wiki/Norman_language
    ["Northern Kurdish"] = 'ku', -- from https://en.wikipedia.org/wiki/Kurmanji
    ["Northern Sami"] = 'se',
    ["Norwegian Bokmål"] = 'nob', -- from https://www.loc.gov/standards/iso639-2/php/code_list.php
    ["Norwegian Nynorsk"] = 'nno', -- from https://www.loc.gov/standards/iso639-2/php/code_list.php
    Occitan = 'oc',
    ["Old Armenian"] = 'xcl', -- picked classical armenian
    ["Old Church Slavonic"] = 'cu',
    ["Old English"] = 'ang', -- from https://www.loc.gov/standards/iso639-2/php/code_list.php
    ["Old French"] = 'fro', -- from https://www.loc.gov/standards/iso639-2/php/code_list.php
    ["Old High German"] = 'goh', -- from https://www.loc.gov/standards/iso639-2/php/code_list.php
    ["Old Irish"] = 'sga', -- from https://www.loc.gov/standards/iso639-2/php/code_list.php
    ["Old Norse"] = 'non', -- from https://www.loc.gov/standards/iso639-2/php/code_list.php
    ["Old Saxon"] = 'nds', -- from https://www.loc.gov/standards/iso639-2/php/code_list.php could be old english too
    Oromo = 'om',
    ["Ottoman Turkish"] = 'ota', -- from https://www.loc.gov/standards/iso639-2/php/code_list.php
    Pali = 'pi',
    Pashto = 'ps',
    ["Pennsylvania German"] = 'pdc', -- from https://en.wikipedia.org/wiki/Pennsylvania_Dutch_language
    Persian = 'fa',
    Phalura = 'phl', -- https://en.wikipedia.org/wiki/Palula_language
    Phuthi = 'bnt', -- it has none https://en.wikipedia.org/wiki/Phuthi_language so i put it under bantu
    Plautdietsch = 'pdt', -- https://en.wikipedia.org/wiki/Plautdietsch_language
    ["Proto-Finnic"] = 'urj', -- doesn't have one so i put it under https://en.wikipedia.org/wiki/Uralic_languages
    ["Proto-Germanic"] = 'gem', -- from https://en.wikipedia.org/wiki/List_of_ISO_639-2_codes
    ["Proto-Samic"] = 'smi', -- from https://en.wikipedia.org/wiki/List_of_ISO_639-2_codes
    ["Proto-Slavic"] = 'sla', -- from https://en.wikipedia.org/wiki/List_of_ISO_639-2_codes
    ["Proto-West Germanic"] = 'gem', -- from https://en.wikipedia.org/wiki/List_of_ISO_639-2_codes
    Punjabi = 'pa',
    Quechua = 'qu',
    Romanian = 'ro',
    Romansch = 'rm',
    Sanskrit = 'sa',
    ["Saterland Frisian"] = 'stq', -- https://en.wikipedia.org/wiki/Saterland_Frisian_language
    Scots = 'sco', -- https://www.loc.gov/standards/iso639-2/php/code_list.php
    ["Scottish Gaelic"] = 'gd',
    ["Serbo-Croatian"] = 'hbs', -- https://en.wikipedia.org/wiki/Serbo-Croatian
    Shan = 'shn', -- https://en.wikipedia.org/wiki/List_of_ISO_639-2_codes
    Sicilian = 'scn', -- https://en.wikipedia.org/wiki/List_of_ISO_639-2_codes
    Slovak = 'sk',
    Slovene = 'sl',
    ["South Levantine Arabic"] = 'apc', -- https://en.wikipedia.org/wiki/South_Levantine_Arabic
    ["Southern Altai"] = 'alt', -- https://en.wikipedia.org/wiki/List_of_ISO_639-2_codes
    Sumerian = 'sux', -- https://en.wikipedia.org/wiki/List_of_ISO_639-2_codes
    Sundanese = 'su',
    Swazi = 'ss',
    Tagalog = 'tl',
    Tajik = 'tg',
    Tamil = 'ta',
    Tangut = 'txgh', -- https://en.wikipedia.org/wiki/Tangut_language
    Tatar = 'tat', -- https://en.wikipedia.org/wiki/List_of_ISO_639-2_codes
    Telugu = 'te',
    Ternate = 'tft', -- https://en.wikipedia.org/wiki/Ternate_language
    Thai = 'th',
    Tibetan = 'bo',
    ["Tocharian B"] = 'ine', -- nothing specific so put it under indo european
    ["Tok Pisin"] = 'tpi', -- from https://en.wikipedia.org/wiki/List_of_ISO_639-2_codes
    Turkish = 'tr',
    Urdu = 'ur',
    Uyghur = 'ug',
    Uzbek = 'uz',
    Venetian = 'vec', -- https://en.wikipedia.org/wiki/Venetian_language
    Veps = 'vep', -- https://en.wikipedia.org/wiki/Veps_language
    Vietnamese = 'vi',
    Vilamovian = 'wym', -- https://en.wikipedia.org/wiki/Wymysorys_language
    ["Volapük"] = 'vo',
    Walloon = 'wa',
    Welsh = 'cy',
    ["West Frisian"] = 'fy',
    Westrobothnian = 'sv', -- can't find anything so put it under swedish
    Xhosa = 'xh',
    Yakut = 'sah', -- https://en.wikipedia.org/wiki/List_of_ISO_639-2_codes
    Yiddish = 'yid', -- https://en.wikipedia.org/wiki/List_of_ISO_639-2_codes
    Yola = 'yol', -- https://en.wikipedia.org/wiki/Yola_language
    Yoruba = 'yo',
    Zhuang = 'za',
    Zulu = 'zu'
}
```

I also took the liberty to change some listed languages which didn't seem correct to me [here](https://github.com/koreader/koreader/pull/10184/commits/039a240372695a0ae0de6350641f5208ed5a98a9)

nairyosangha commented:
Then for actually storing this info in the downloaded dictionary's .ifo file: https://github.com/koreader/koreader/pull/10184/commits/1377fae47be01c51fcb08d9b7624c0e7687efb98

I had to change that extract function, because I wanted to be able to control the actual directory the .ifo ends up in. It seems like this thing is only used in readerdictionary anyway so I hope it's okay.
As an upside, storing the downloaded dictionary under datadir/dict.name looks more neat to me :slightly_smiling_face:

Frenzie commented on frontend/apps/reader/modules/readerdictionary.lua:
```diff
@@ -1118,9 +1123,13 @@ function ReaderDictionary:downloadDictionary(dict, download_location, continue)
 
-    local ok, error = Device:unpackArchive(download_location, self.data_dir)
+    -- stable target directory is needed so we can look through the folder later
+    local dict_path = self.data_dir .. '/' .. dict.name
```
Very much a nit, but please stick to the double quotes style unless there's a good reason to deviate.

Frenzie commented:
> local mapper = {

This seems to be a bit backwards? :-) I forget the exact rationale from back in the day, but if you're including info like `de-ru` it presumably makes more sense to use that as the source, possibly duplicated in the data structure if that makes things faster.

nairyosangha commented:
> > local mapper = {
> 
> This seems to be a bit backwards? :-) I forget the exact rationale from back in the day, but if you're including info like `de-ru` it presumably makes more sense to use that as the source, possibly duplicated in the data structure if that makes things faster.

That stuff I posted isn't actually in the code, it's just something I used locally to extend that dictionaries.lua file. I figured it'd be easier to look over that since the languages aren't duplicated in that list, easier than skimming over 3k+ lines

NiLuJe commented:
Don't we have a mapping somewhere for font and/or harfbuzz stuff? (e.g., xtext or the font cache thingy?).

Frenzie commented:
@nairyosangha 
It's a generic design remark, though as far as this PR goes the only question is whether it makes sense to have a separate         `ifo_lang` or if it should be constructed dynamically out of `lang_in` and `lang_out`. But having that `ifo_lang` there flaunting language codes makes the original design look a bit silly or worse, and your correction from Deutsch to German for consistency shows that I probably didn't make the right choice when I first wrote it.

Given that I can't imagine extendIfoWithLanguage() being called very often there's no design/speed argument in favor of including it explicitly, though there's no very strong argument against it either.

@NiLuJe 
> Don't we have a mapping somewhere for font and/or harfbuzz stuff?

Yes, in frontend/language or something, unless you mean another one. There's also a Google Translate API somewhere.

poire-z commented on frontend/apps/reader/modules/readerdictionary.lua:
```diff
@@ -1118,9 +1123,13 @@ function ReaderDictionary:downloadDictionary(dict, download_location, continue)
         return false
     end
 
-    local ok, error = Device:unpackArchive(download_location, self.data_dir)
+    -- stable target directory is needed so we can look through the folder later
+    local dict_path = self.data_dir .. "/" .. dict.name
+    util.makePath(dict_path)
+    local ok, error = Device:unpackArchive(download_location, dict_path, true)
```
Dunno how problematic or not it is to make a directory named as the name= of what's in our dictionaries.lua:
- with their possible non-ASCII content - but I guess everybody is already able to create file and directories with books on all devices
- the fact that it seems we have 7 duplicated names - but then, before, just untaring and keeping the subdirectories in the downloaded tar.gz, we could already have duplicated directories.

nairyosangha replied:
I figured it should be okay since there's nothing preventing both of those things with the current code either.
Maybe we could make those dict names in the list unique? Not sure if it's worth the effort.

poire-z commented on frontend/apps/reader/modules/readerdictionary.lua:
```diff
@@ -1135,6 +1144,24 @@ function ReaderDictionary:downloadDictionary(dict, download_location, continue)
     end
 end
 
+function ReaderDictionary:extendIfoWithLanguage(dict, download_location)
+    local function cb(path, filename)
+        if util.getFileNameSuffix(filename) == "ifo" and dict.ifo_lang then
```
(Dunno if in the future we will want to add other things than ifo_lang - but you check dict_ifo_lang last - if you checked it first on line 1132, you could avoid calling extendIfoWithLanguage() when it is not set. But ok, it's now always set :)

nairyosangha replied:
good point, I did call it extendIfo**WithLanguage** specifically so I might as well check it up front. :smile:

pazos commented on frontend/device/generic/device.lua:
```diff
@@ -920,8 +921,12 @@ function Device:unpackArchive(archive, extract_to)
 
-function Device:untar(archive, extract_to)
-    return os.execute(("./tar xf %q -C %q"):format(archive, extract_to))
+function Device:untar(archive, extract_to, with_stripped_root)
```
This function is also in https://github.com/koreader/koreader/blob/master/frontend/device/android/device.lua#L514.

Not sure what '--strip-components=1' does but we need to check how android behaves without any additional change with this PR in and/or change the untar logic in https://github.com/koreader/android-luajit-launcher/blob/master/app/src/main/java/org/koreader/launcher/extensions/FileExtensions.kt#L74-L126

pazos replied:
Or leave this function signature as is and do "strip components" by hand after the tar file is uncompressed

nairyosangha replied:
I don't think it will cause any issues, the `--strip-components=1` will remove the root directory in the archive (if it's present).
So when we pass in the `extract_to` <data_dir>/<dict.name> directory it doesn't create an unnecessary nested folder inside that deepest directory. 

On android there would be an unnecessary nested directory, but it wouldn't fail right? It looks like it'll make the `extract_dir` just fine, I assume it can create parent directories fine as well?

pazos replied:
Thanks for the info.

> On android there would be an unnecessary nested directory, but it wouldn't fail right? 

Nope, it should work fine. Still I would prefeer to keep the same expected behaviour but I think it doesn't need to be addressed in this PR.


>  the --strip-components=1 will remove the root directory in the archive (if it's present).

Do you have some samples of dicts with and without an inner folder?

nairyosangha replied:
I couldn't immediately find one so I made one myself for a random small one of the list: https://github.com/Vuizur/Wiktionary-Dictionaries/raw/master/Alemannic%20German-English%20Wiktionary%20dictionary%20stardict.tar.gz
```sh
$ tar tf Alemannic%20German-English%20Wiktionary%20dictionary%20stardict.tar.gz 
Alemannic German-English Wiktionary dictionary stardict/
Alemannic German-English Wiktionary dictionary stardict/Alemannic German-English Wiktionary dictionary.dict.dz
Alemannic German-English Wiktionary dictionary stardict/Alemannic German-English Wiktionary dictionary.idx
Alemannic German-English Wiktionary dictionary stardict/Alemannic German-English Wiktionary dictionary.ifo
Alemannic German-English Wiktionary dictionary stardict/Alemannic German-English Wiktionary dictionary.syn
$ # this one clearly has a root directory
$ tar cfz german-english-noroot.tar.gz -C Alemannic\ German-English\ Wiktionary/ 'Alemannic German-English Wiktionary dictionary.'{dict.dz,ifo,idx,syn}
$ tar tf german-english-noroot.tar.gz 
Alemannic German-English Wiktionary dictionary.dict.dz
Alemannic German-English Wiktionary dictionary.ifo
Alemannic German-English Wiktionary dictionary.idx
Alemannic German-English Wiktionary dictionary.syn
$ # root's gone!
```
[german-english-noroot.tar.gz](https://github.com/koreader/koreader/files/10931594/german-english-noroot.tar.gz)

nairyosangha commented:
> It's a generic design remark, though as far as this PR goes the only question is whether it makes sense to have a separate `ifo_lang` or if it should be constructed dynamically out of `lang_in` and `lang_out`. But having that `ifo_lang` there flaunting language codes makes the original design look a bit silly or worse, and your correction from Deutsch to German for consistency shows that I probably didn't make the right choice when I first wrote it.

I see what you mean now! I agree with that, that's why I tried using those existing lists in the code first ( https://github.com/koreader/koreader/pull/10184#issuecomment-1458579784 ), but neither of those seemed like great candidates.

Maybe `lang_in` and `lang_out` could refer to the iso2 codes used in `ifo_lang` now, and then those could be converted to the full name (maybe even localized?) before inserting them in the download menu?
I would need to add that list I made somewhere in the code then though, or figure out if there's a proper one hiding in the code somewhere :slightly_smiling_face:

poire-z commented:
I wouldn't bother more than you did, now that you have added the ifo_lang tuple to all of them.

(There is another such list here:)
https://github.com/koreader/koreader/blob/a3d42ac3444f78e354d7aad2926a476f29813fe5/frontend/ui/translator.lua#L24-L31

Frenzie commented:
> (maybe even localized?)

Not maybe, that's a given. ;-)

> I wouldn't bother more than you did, now that you have added the ifo_lang tuple to all of them.

Hold on, I'm not at all convinced that's a desirable thing to do.

nairyosangha commented:
I gave it a go after all :slightly_smiling_face: 
Since that biggest list in `translator.lua` still isn't nearly complete, and since it's not supposed to be a full list anyway, I just made my own list.

I think it's neat that the same info isn't stored twice in `dictionaries.lua` now, and having the languages localized in the menu is nice too. However, not sure it's a good idea to keep yet another list. I guess it won't hurt? I can always revert the change

poire-z commented on frontend/ui/isolanguage.lua:
```diff
@@ -0,0 +1,530 @@
+
+function IsoLanguage:getLocalizedLanguage(lang_locale)
+    local language = locale_lang_map[lang_locale]
+    return language and _(language) or lang_locale
```
This won't propose any of your language string to translators (and I don't think we want to propose them :) It will just benefit from the ones that may be translater from elsewhere I guess.
So, maybe just don't localize them ?
@Frenzie ?

Frenzie replied:
It wouldn't make too much sense to preemptively localize (particularly the likes of Blissymbols), but note that this won't localize much as written, except insofar as the same string is coincidentally localized elsewhere already. :-) The localization would need to happen in isolanguage.lua.

Concretely, only strings that look like _("") or _([[]]) are picked up for localization.

poire-z commented on frontend/apps/reader/modules/readerdictionary.lua:
```diff
@@ -7,6 +7,7 @@ local Event = require("ui/event")
 local Geom = require("ui/geometry")
 local InfoMessage = require("ui/widget/infomessage")
 local InputDialog = require("ui/widget/inputdialog")
+local IsoLanguage = require("ui/isolanguage")
 local JSON = require("json")
```
You'd better require() it only in the function where you need it. We don't want all these language name strings to say in memory each time KOreader is launched - but only for the rare session this menu will be used.

poire-z commented on frontend/apps/reader/modules/readerdictionary.lua:
```diff
@@ -480,12 +479,13 @@ end
 
 function ReaderDictionary:_genDownloadDictionariesMenu()
     local downloadable_dicts = require("ui/data/dictionaries")
+    local isoLanguage = require("ui/isolanguage")
```
(You can keep its Uppercase Initial `I` :) It's not the fact that it's now no longer a module level require that should make it lose its uppercase - we have many `local ConfirmBox = require("ui/widget/confirmbox")` - and `local confirmbox = ConfirmBox:new{}` in functions - it's rather if it is a class or not, `downloadable_dicts` is probably just a table, but your `IsoLanguage`, having methods, probably deserves being called `local IsoLanguage`.)

nairyosangha replied:
Okay, I'll change it back
Shouldn't that socket stuff lower down follow that same logic then? That's what I looked at, I figured I should follow that

Frenzie commented on frontend/ui/isolanguage.lua:
```diff
@@ -0,0 +1,530 @@
+    am = "Amharic",
+    ar = "Arabic",
+    an = "Aragonese",
+    hy = "Armenian",
```
I won't ask you to change it unless you want to or agree with me, but I find the English order a bit confusing here.

nairyosangha commented:
Updated the `dictionaries.lua` file and the reference list yet again, using iso-3 everywhere this time, alphabetically sorted (per chunk).
For the ones where there are accompanying iso-2 codes, I've kept all of them in there. For the rest, only the stuff that's actually used is present, which cuts it down quite a bit from before.

I removed some stuff from the 'official' name where it seemed appropriate and pasted it after it as a comment.
There's also still a few entries right at the end which just don't seem to have a proper iso-3 code for them, so I picked the overarching iso code, which doesn't look great in the menu, but I'm not sure what to put in there otherwise.

I also moved the reference list to the same folder as dictionaries.lua, since it's kinda data as well

nairyosangha commented:
I'm also wondering if `cmn` for mandarin and `yue` for cantonese are appropriate, I've seen that represented as `zh_CH` and `zh_TW` in some places in the code

poire-z commented:
I haven't checked, but remember these lang codes (thanks to this PR) are given to Harfbuzz to possibly find them specified in fonts. I think HB `may` support the 3 chars lang code (and eventually find the equivalent 2-chars lang code), but it was may be safer to keep using the most common 2 chars lang code.

nairyosangha commented:
> I haven't checked, but remember these lang codes (thanks to this PR) are given to Harfbuzz to possibly find them specified in fonts. I think HB `may` support the 3 chars lang code (and eventually find the equivalent 2-chars lang code), but it was may be safer to keep using the most common 2 chars lang code.

Uh oh, seems like you're right, putting `lang=jpn-jpn` in my dicts doesn't work, only the iso2 code does :sweat_smile:

poire-z commented:
Harfbuzz expects "bcp47" it seems.
https://github.com/harfbuzz/harfbuzz/blob/main/src/hb-ot-tag-table.hh
https://github.com/harfbuzz/harfbuzz/blob/main/src/gen-tag-table.py

I think it has a table mapping them to the Opentype language, and the 3-char code for Japanese seems to be JAN rather than JPN. Are you sure it's really jpn in your ISO list?

poire-z commented:
> However, even though I create that ifo_lang thing BEFORE I convert lang_in and lang_out, it functions as if I didn't. I end up with it being something like "English-English" instead of the "eng-eng" I would expect.

May be only on 2nd++ iteration of it (if this is a thing thay happens dunno) ? You get the dict object from downloadable_dicts , you upadte its ifo_lang, then you update their lang_in. On the next iteration, you'll find lang_in as you fixed them, and screw your original ifo_lang ?

nairyosangha commented:
> Harfbuzz expects "bcp47" it seems. https://github.com/harfbuzz/harfbuzz/blob/main/src/hb-ot-tag-table.hh https://github.com/harfbuzz/harfbuzz/blob/main/src/gen-tag-table.py
> 
> I think it has a table mapping them to the Opentype language, and the 3-char code for Japanese seems to be JAN rather than JPN. Are you sure it's really jpn in your ISO list?

looks like it..
https://iso639-3.sil.org/code/jpn
https://iso639-3.sil.org/code/jan

nairyosangha commented:
> > However, even though I create that ifo_lang thing BEFORE I convert lang_in and lang_out, it functions as if I didn't. I end up with it being something like "English-English" instead of the "eng-eng" I would expect.
> 
> May be only on 2nd++ iteration of it (if this is a thing thay happens dunno) ? You get the dict object from downloadable_dicts , you upadte its ifo_lang, then you update their lang_in. On the next iteration, you'll find lang_in as you fixed them, and screw your original ifo_lang ?

that's exactly it, apparently it already loads that stuff the moment the menu is shown, seems like it's called twice before I've even clicked the button to show the downloadable dicts :sweat_smile: , so I'll just have to put a guard around it I guess

poire-z commented:
> loads that stuff the moment the menu is shown

The top menu ?!! Or the Download dict submenu ?
If the former, we probably should switch some bits to sub_item_table_func or somethingelse_func, so it gets loaded only when the menu is entered.

nairyosangha commented:
> > loads that stuff the moment the menu is shown
> 
> The top menu ?!! Or the Download dict submenu ? If the former, we probably should switch some bits to sub_item_table_func or somethingelse_func, so it gets loaded only when the menu is entered.

I've put a logger statement on top of that `_genDownloadDictionariesMenu` that just shows me the first dict. 
The moment I click the top of the screen (so the main menu pops up), it prints that first dict twice.

It doesn't happen again even when navigating to the actual menu that shows me all the dictionaries though. Not sure why it has to do it twice

poire-z commented:
Can you test if this makes it load only when the submenu is entered ?
```diff
--- a/frontend/apps/reader/modules/readerdictionary.lua
+++ b/frontend/apps/reader/modules/readerdictionary.lua
@@ -265,7 +265,7 @@ function ReaderDictionary:addToMainMenu(menu_items)
             },
             {
                 text = _("Download dictionaries"),
-                sub_item_table = self:_genDownloadDictionariesMenu()
+                sub_item_table_func = function() return self:_genDownloadDictionariesMenu() end,
             },
             {
                 text = _("Enable fuzzy search"),
```
just so we don't get to load and keep all the dictionaries.lua strings in memory.

poire-z commented:
(Sorry for the close/open.)
The above doesn't change the fact that you indeed have to set your ifo_lang only once :)

nairyosangha commented:
> Can you test if this makes it load only when the submenu is entered ?
> 
> ```diff
> --- a/frontend/apps/reader/modules/readerdictionary.lua
> +++ b/frontend/apps/reader/modules/readerdictionary.lua
> @@ -265,7 +265,7 @@ function ReaderDictionary:addToMainMenu(menu_items)
>              },
>              {
>                  text = _("Download dictionaries"),
> -                sub_item_table = self:_genDownloadDictionariesMenu()
> +                sub_item_table_func = function() return self:_genDownloadDictionariesMenu() end,
>              },
>              {
>                  text = _("Enable fuzzy search"),
> ```
> 
> just so we don't get to load and keep all the dictionaries.lua strings in memory.

Ya that fixes it, only happens once now, and only when I actually click the button to show me the downloadable dictionaries

nairyosangha commented:
> Harfbuzz expects "bcp47" it seems. https://github.com/harfbuzz/harfbuzz/blob/main/src/hb-ot-tag-table.hh https://github.com/harfbuzz/harfbuzz/blob/main/src/gen-tag-table.py
> 
> I think it has a table mapping them to the Opentype language, and the 3-char code for Japanese seems to be JAN rather than JPN. Are you sure it's really jpn in your ISO list?

reading on wikipedia, the language subtag should be:
> A single primary language subtag based on a two-letter language code from [ISO 639-1](https://en.wikipedia.org/wiki/ISO_639-1) (2002) or a three-letter code from [ISO 639-2](https://en.wikipedia.org/wiki/ISO_639-2) (1998), [ISO 639-3](https://en.wikipedia.org/wiki/ISO_639-3) (2007) or ISO 639-5 (2008), or registered through the BCP 47 process and composed of five to eight letters;

So, that first chunk in isolanguage.lua (which does have an iso2) actually needs to be changed back to iso2 codes then? Or is there a way to get around it

nairyosangha commented:
I could of course also add another local mapping in that file which maps the iso3 to iso2 when possible, not sure if that's any better than just having iso2 codes in the main mapping directly

poire-z commented:
Haven't really followed all your languages code investigations/experiments
But... do simple, do only what's needed :) (ie. no need to have some iso-5H1T if we won't need it anywhere else, or we might need iso-C54P in other contexts so your iso-5H1T won't help - and only add the codes we actually have dicts for them, we're not building a reference, we only need strings for what we support.)

Frenzie commented:
> So, that first chunk in isolanguage.lua (which does have an iso2) actually needs to be changed back to iso2 codes then? Or is there a way to get around it

If there's a technical reason that's all well and good. It's just a bit confusing to look at. But how does that affect those relatively minor languages?

nairyosangha commented:
> > So, that first chunk in isolanguage.lua (which does have an iso2) actually needs to be changed back to iso2 codes then? Or is there a way to get around it
> 
> If there's a technical reason that's all well and good. It's just a bit confusing to look at. But how does that affect those relatively minor languages?

You mean the ones which don't have an iso2 code equivalent? According to that BCP47 standard you're supposed to use the iso3 code in those cases. So I guess it should just work. Although I don't really know how to test it properly.
It definitely seems like it's not gonna complain if it gets a language code it doesn't know, it'll just look like it already does currently I suppose. That's what happened when I tested `lang=jpn-jpn` anyway.

The annoying thing is that you'll sometimes have iso3 codes, sometimes iso2 codes in that `dictionaries.lua` file, which is a bit unfortunate but probably acceptable?

Another approach would just be to go back to @poire-z 's initial suggestion, and add the `ifo_lang` to some of the dictionaries in `dictionaries.lua` where we KNOW it would help, and leave all the stuff where it won't help alone, like everything in standard latin script, or the strange esoteric languages. Sure you'll have some info duplicated, but you'll need way fewer changes.

I'm up for both :smile:

Frenzie commented:
> It definitely seems like it's not gonna complain if it gets a language code it doesn't know, it'll just look like it already does currently I suppose. That's what happened when I tested lang=jpn-jpn anyway.

Right, but that's what I mean, will the text shaping on some Indic language still work or does that need some kind of workaround?

> The annoying thing is that you'll sometimes have iso3 codes, sometimes iso2 codes in that dictionaries.lua file, which is a bit unfortunate but probably acceptable?

In the end what's acceptable is what works. ;-)

poire-z commented:
The current issue is about the dictionaries we provide for download - for a strange indic language dict a user added himself, he will have to manually add to its .ifo the lang= tag, and he will have to decide/test what values to use. What he sets is just for Harfbuzz needs (and what the added fallback font Noto Indicish supports), and won't impact our UI dict list - so we don't need to know and have that language in any isolanguage.lua.
If someone adds a new dict to our download dicts list, he would add the appropriate ifo_lang= - it's possible he may need this, but won't need to add a new language to the download dict language list - so it may be that the lang_in/lang_out=English can be orthogonal to the rendering language ifo_lang=oang (old anglo?).
So, I'd say: do simple, do flexible, don't assume we're clever :):

> add the ifo_lang to some of the dictionaries in dictionaries.lua where we KNOW it would help

nairyosangha commented:
> What he sets is just for Harfbuzz needs (and what the added fallback font Noto Indicish supports), and won't impact our UI dict list - so we don't need to know and have that language in any isolanguage.lua.

Agreed, and that's what I did now. Well, I guess the first chunk of iso639-1 codes can probably be even smaller, I don't think they're all used either. But for the rest, only stuff we actually need is listed.

> If someone adds a new dict to our download dicts list, he would add the appropriate ifo_lang= - it's possible he may need this, but won't need to add a new language to the download dict language list - so it may be that the lang_in/lang_out=English can be orthogonal to the rendering language ifo_lang=oang (old anglo?).

So you're suggesting to keep the `ifo_lang` field in `dictionaries.lua` separate from `lang_in` / `lang_out`?
Because if not, it's definitely possible that the user _would_ have to extend the `isolanguage.lua` list as well.

@Frenzie indicated [earlier](https://github.com/koreader/koreader/pull/10184#issuecomment-1460511051) that he wasn't convinced that's a good idea though, unless I misunderstood

I'll just reiterate that for me it doesn't really matter, I'd be happy just being able to manually add an `ifo_lang` field to a dictionary myself, be it one I added, or one that was downloaded by KOreader :slightly_smiling_face:

poire-z commented:
> So you're suggesting to keep the ifo_lang field in dictionaries.lua separate from lang_in / lang_out?

Yes, that's what I'm suggesting.




> @Frenzie indicated [earlier](https://github.com/koreader/koreader/pull/10184#issuecomment-1460511051) that he wasn't convinced that's a good idea though, unless I misunderstood

Not sure what he had in mind, but after all our dicussions, the various specs about language codes, the need to have the type of code supported by Harfbuzz, the non-real need to have a proper full list and translations, the fact that having them orthogonal allowing more flexibility... may be he can revamp his thoughts :)

Frenzie commented:
> @Frenzie indicated https://github.com/koreader/koreader/pull/10184#issuecomment-1460511051 that he wasn't convinced that's a good idea though, unless I misunderstood

There are plenty of scenarios in which it could make sense. for example if one of them is for Harfbuzz and the other something completely different for the string in the UI. But earlier it was just a duplicate.

But what I meant was to provide me with a rationale, the reasoning behind the decision. ;-)

nairyosangha commented:
I think there's 3 options to consider

### minimal solution:
- Only add the functionality to read in `ifo_lang` from a local dictionary. That would cover my use case already, and for the downloaded dictionaries, there's nothing stopping the users from adding it themselves.
   + Fewer changes, no messing around with language codes
   -  Arguably incomplete experience when using downloaded dictionaries

I'm leaning towards this because I do not really need nor use the provided download dictionaries option. You don't have any duplicate info in `dictionaries.lua`. There's also no chance of anything in a dict being rendered incorrectly because of this PR, since all these changes have to be made by the user themselves.

### simple solution:
- Allow adding an extra `ifo_lang` to the dict where applicable to improve text rendering where I know it would help. I really can only tell myself that it helps for japanese, so we could add it only for those few entries.
  + Some of the downloaded dictionaries will immediately look better (still not perfect since we can't handle html dicts)
  + only a few changes need to be made to `dictionaries.lua`, we're not duplicating tons of info.

I think the duplication of info in `dictionaries.lua` would still be acceptable here. We know this issue is mainly for CJK languages, so you barely need to add any info.

### 'complicated' solution:
- Change all `lang_in` & `lang_out` entries in the dict to some iso code
   + either `dictionaries.lua` contains only iso3 codes, or iso2 when they exists, if not, iso3, could be made clear with a comment on top
- Provide mapping to convert between `lang_in`/`lang_out` to harfbuzz format and to full language name.
   + mapping to harfbuzz format wouldn't be necessary if `dictionaries.lua` contains that already

I don't want to have to check if every non-latin language is still rendered correctly, I wouldn't know what to look for and it would take forever. There's also some dictionaries which just don't map to a proper language (like proto-anything), which means you have to use whatever macrolanguage is closest, which means the full name won't like nice in the UI.



What it comes down to is that I really just wanna be told what's best, and then I'll go with that :sweat_smile: 
As long as the simple solution is in place, my problem is solved, the rest is just extra

poire-z commented:
Simple solution is nice.
I don't think we need yet another isosomething mapping (we already have a few, our ui languages and xtext, google code for translator, typography language for crengine), no need for another one, when we don't yet know of any feature that would _require_ it to be iso2 or iso3, and we then can't thinkg think about additional mappings for protogerman.
No need for a protofuturekoreader, until we know into what it should evolve :)

nairyosangha commented:
> The third is only complicated because you wrote it complicated. I'd just make it simple and simple.
> 
> * `ifo_lang` for HB (not sure if the format you initially picked still makes sense?) — although I'm not opposed to a mapping
> * `lang_in` and `lang_out` in ISO-_mumble_
> * some custom extensions for ISO for Proto-Germanic and friends (something like `pge` or `pgr`, or perhaps something like `xpg` because that looks more custom)

but isn't having both `ifo_lang` AND `lang_in/out` in some iso format also duplication of info? Idk how that's any better than having the full language names, plus `ifo_lang` separately, they still represent the same thing, which is what I thought you were not a fan of

nairyosangha commented:
> It looked like unnecessary duplication. That's a fairly big difference. ;-) But I don't think anyone provided an argument in favor of doing it that way? The only argument I recall seeing (HarfBuzz) seems to have presented itself by accident.

Isn't not having to mess with mappings a pretty big argument in its favor? :smiley: 

```
{
        name = "JMdict Japanese-English dictionary",
        lang_in = "jpn",
        lang_out = "eng",
        ifo_lang = "ja-en",
        entries = 188380,
        license = "CC-BY-SA 3.0 (from https://www.edrdg.org/wiki/index.php/JMdictDB_Project)",
        url = "https://cyphar.github.io/jpn-stardicts/JMdict-ja-en.tar.gz"
 }
```
I agree that it would be nice to have something like this, but then you might as well have a mapping to determine `ifo_lang` based on the other two as well.

How about going for the simple solution in this PR, and doing the `lang_in` / `lang_out` in a separate one? Since the point of doing that seems more for localization purposes than for the correct glyphs.

Frenzie commented:
> Isn't not having to mess with mappings a pretty big argument in its favor? smiley

You need a mapping to easily ensure that German isn't Deutsch or vice versa. So no. ;-)

> I agree that it would be nice to have something like this, but then you might as well have a mapping to determine ifo_lang based on the other two as well.

You're the one who called it complicated, lol. I was just trying to compromise. It sounds like the way to go as far as I'm concerned.

poire-z commented:
I wrote above:

> so it may be that the lang_in/lang_out=English can be orthogonal to the rendering language ifo_lang=oang (old anglo?).

So, if we assume it may happen that we need non-generated-from-a-map-from lang_in/lang_out values for ifo_lang, and/or want to allow some flexibility in there, you don't need any new iso2/iso3: just wrap all natural english language names in `_()` so they can get translated :)

(And translators will thank us for having only a little set of languages to translate - the ones we have dictionaries to download for - and not hundreds fo them we will never use if coming from a iso2/3 full list :)

Frenzie commented:
That's fine too.

nairyosangha commented:
> > I agree that it would be nice to have something like this, but then you might as well have a mapping to determine ifo_lang based on the other two as well.
> 
> You're the one who called it complicated, lol. I was just trying to compromise. It sounds like the way to go as far as I'm concerned.

In my defense, I called it 'complicated' because I already had a simple solution :smile: 

I think I have what you want now, `dictionaries.lua` uses iso3 everywhere. I made up some fake ones for the proto languages. `isolanguage.lua` contains only the iso3 codes we actually use. There's also a separate mapping in there to convert to BCP47, which is used to build `ifo_lang` dynamically.

> (And translators will thank us for having only a little set of languages to translate - the ones we have dictionaries to download for - and not hundreds fo them we will never use if coming from a iso2/3 full list :)

It's still quite a list now, even with only the stuff that's actually used :smiley:

Frenzie commented on frontend/ui/data/dictionaries.lua:
```diff
@@ -4,2857 +4,2857 @@ local _ = require("gettext")
         url = "https://gitlab.com/avsej/dicts-stardict-form-xdxf/raw/d636cc5e8d4a47e22ac7466f4af6d435a8a3f650/002c/stardict-comn_sdict05_dutch-english-2.4.2.tar.gz"
     },
     {
         name = "Engligh Idioms (eng-eng)",
```
Not really part of this PR, but maybe while you're in here?

Frenzie commented on frontend/ui/data/dictionaries.lua:
```diff
@@ -4,2857 +4,2857 @@ local _ = require("gettext")
         url = "https://gitlab.com/avsej/dicts-stardict-form-xdxf/raw/d636cc5e8d4a47e22ac7466f4af6d435a8a3f650/002c/stardict-comn_sdict05_idioms_eng_eng-2.4.2.tar.gz"
     },
     {
         name = "Engligh Idioms (eng-rus)",
```
Same

Frenzie commented on frontend/ui/data/isolanguage.lua:
```diff
@@ -0,0 +1,339 @@
+local locale_lang_map = {
+    -- from https://iso639-3.sil.org/sites/iso639-3/files/downloads/iso-639-3.tab
+    -- first chunk of names are languages present in iso-639-1
+    aar = "Afar",
```
If we want them to be translated they'd all need to be wrapped in _().
```suggestion
    aar = _("Afar"),
```

nairyosangha replied:
Should we do it for all of them, or only the first (most common) chunk?

Frenzie replied:
Who cares about those, it's the Proto-Germanic that we really care about. ;-P

I'd just do all, otherwise I can all but guarantee someone will figure it out eventually. On the plus side, I suspect there's a decent overlap with the already translated Google Translate list.

poire-z replied:
Better add @translators notes, if you don't want to get `Afar` translated to `De loin`...
Poor translators :/ and our translations % will drop, nobody sane will translate all these languages...

Frenzie replied:
Speaking of translator's notes, all of these have been pre-translated already for many languages, so they will show up as suggestions in the Hosted Weblate global memory.

https://hosted.weblate.org/projects/iso-codes/iso-639-2/

So while it will look a bit silly, the best course of action may be to precede every line with a translator's comment like:
```lua
-- @translators Most of these language name have already been translated at <https://hosted.weblate.org/projects/iso-codes/iso-639-2/>. Click "Automatic suggestions" to see them below the textfield.
```

Frenzie replied:
Or at least on the first one here, which is generally the one translators will see first.

Frenzie reviewed (approved):
I'm basically happy with this. :+1:

poire-z commented:
Too much uneeded stuff for my liking, but if @Frenzie is happy, well, nothing against :)

Frenzie commented on frontend/ui/data/isolanguage.lua:
```diff
@@ -0,0 +1,339 @@
+    yue = _("Yue Chinese"),
+
+    -- these are made up language codes for very specific languages which do not map to a specific ISO3 code
+    xxpf = _("Proto Finnic"),
+    xxpg = _("Proto Germanic"),
+    xxpw = _("Proto West-Germanic"),
+    xxps = _("Proto Sami"),
+    xxsl = _("Proto Slavic"),
```
They're generally spelled more like this, cf. <https://www.britannica.com/topic/Indo-European-languages/Characteristic-developments-of-Indo-European-languages#ref603329>.
```suggestion
    xxpf = _("Proto-Finnic"),
    xxpg = _("Proto-Germanic"),
    xxpw = _("Proto-West Germanic"),
    xxps = _("Proto-Sami"),
    xxsl = _("Proto-Slavic"),
```

Though I'll agree that Proto-West Germanic looks a bit odd. ^_^

poire-z commented:
@Frenzie : good to go?
(will need a proper summary in the commit message)
